### PR TITLE
MouseMap: custom container widget to map the device SVG

### PR DESCRIPTION
**Pull request by Hjdskes**
This widget implements a central part of the new UI. See [this webm](https://hjdskes.github.io/img/blog/gsoc-part-4/mousemap-interactive.webm) to get an impression for that it does (note that the styling of the highlight is outdated; it now highlights the entire button instead of just its border).

Next up will be a new window making use of this widget, stealing code from piper.py.

### CODE REVIEW

**Comment by Hjdskes** [comment]
If you agree with this approach I can finalize the SVG guidelines in libratbag, and ask Jakub to do some more devices for us.

**Comment by Hjdskes** [comment]
I just noticed that there is a 1 or 2 pixel offset with the right-aligned buttons. I think this is a rounding error but I'm not sure where it comes from.

For example, with a spacing and border width of 30, here is how this should align:

Left aligned buttons: x=30, width=95, sum=125
Spacing: x=125, width=30, sum=155
SVG: x=155, width=468, sum=623
Spacing: x=623, width=30, sum=653
Now, right aligned children should have x=653 but they have 651 instead.

This also happens for SVGs without left-aligned children, so I suspect it has something to do with librsvg. According to the documentation, it returns ints but if it does some rounding internally that could cause the offset. I'm going to look into this now but wanted to post this first so you won't merge it before I'm done.

**Comment by Hjdskes** [comment]
To be explicit about types I rounded the returned floats to their nearest integer in `_translate_to_origin`, but I cannot fix the 1 or 2 pixel offset due to rounding. It's only visible with no spacing and no border width; the more of those the less visible it gets.

File: piper/mousemap.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        self._device = ratbagd_device
+        self._children = []
+        self._highlight_element = None
+
+        self._handle = Rsvg.Handle.new_from_file(ratbagd_device.svg_path)
+        self._svg_data = etree.parse(ratbagd_device.svg_path)
+
+    def do_add(self, widget):
+        """Not implemented, use `add(widget, svg_id)` instead."""
+        pass
```
Comment (whot) [inline piper/mousemap.py]: should we throw an exception here? seems like a programming error if that is ever called

File: piper/mousemap.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        width, the natural child widths (left and right), spacing and border
+        width.
+        """
+        width = 2 * self.props.border_width
+        width_svg = self._handle.props.width
+        width_left = max((child.widget.get_preferred_width()[1] for
+                          child in self._children if child.is_left), default=0)
+        width_right = max((child.widget.get_preferred_width()[1] for
+                          child in self._children if not child.is_left),
+                          default=0)
```
Comment (whot) [inline piper/mousemap.py]: nitpick: split the list comprehension up and make this two lines each, it'll be a lot easier to read if you have a
````
left_children = [foo for foo in foo]
width_left = max(left_children)
````

File: piper/mousemap.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            'sodipodi': 'http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd',
+            'cc': 'http://web.resource.org/cc/',
+            'svg': 'http://www.w3.org/2000/svg',
+            'dc': 'http://purl.org/dc/elements/1.1/',
+            'xlink': 'http://www.w3.org/1999/xlink',
+            'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
+            'inkscape': 'http://www.inkscape.org/namespaces/inkscape'
+        }
+        query = "//svg:rect[@id=\"{}\"][contains(@style, \"{}\")]".\
+                format(svg_id, style)
```
Comment (whot) [inline piper/mousemap.py]: screw the 78 char limit here and make this one line please, much more obvious that way

File: piper/mousemap.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    def _translate_to_origin(self):
+        # Translates the coordinate system such that the SVG and its buttons
+        # will be drawn in the center of the allocated space. The returned x-
+        # and y-coordinates will be the top left corner of the centered SVG.
+        allocation = self.get_allocation()
+        width = self.get_preferred_width()[1]
+        height = self.get_preferred_height()[1]
+
+        width_left = max((child.widget.get_preferred_width()[1] for
+                          child in self._children if child.is_left), default=0)
```
Comment (whot) [inline piper/mousemap.py]: same here, please split it over two lines

**Comment by whot** [comment]
Just a few nitpicks, please push when they're addressed

File: piper/mousemap.py (hunk 1 above)
Comment (Hjdskes) [inline piper/mousemap.py]: Perhaps it is better to simply not implement this virtual/interface method at all? In that case, it won't be listed in the documentation (can also do this by changing the docstring to a comment), and if someone tries to call this method on our container anyway they'll fall through to https://github.com/GNOME/gtk/blob/master/gtk/gtkcontainer.c#L1575

Comment (Hjdskes) [inline piper/mousemap.py]: In fact, Python will automatically reference the `add(widget, svg_id)` method: 

```
 > ./piper.in
Running from source tree, using local files
Traceback (most recent call last):
  File "/home/jente/code/src/github.com/libratbag/piper/piper/application.py", line 53,
in do_activate
    window = Window(self, self._ratbag)
  File "/home/jente/code/src/github.com/libratbag/piper/piper/window.py", line 102, in _
_init__
    stack.add_titled(self._setup_buttons_page(device), "buttons", _("Buttons"))
  File "/home/jente/code/src/github.com/libratbag/piper/piper/window.py", line 145, in _
setup_buttons_page
    mousemap.add(mapbutton)
TypeError: add() missing 1 required positional argument: 'svg_id'
```

To use the virtual `do_add` method and fall through to the line in my earlier comment, the user (developer) will have to explicitly call `mousemap.do_add(mousemap, widget)` which is unlikely to happen.

So I'll just remove `do_add` and we should be fine.

**Comment by Hjdskes** [comment]
@whot (and others): I just pushed a skeleton window that makes it easier to test the changes. Should you want to test them, make sure you download [these SVGs](https://gist.github.com/Hjdskes/b96b351ae0920d56959de084d45d53fa) that have been modified according to the [new guidelines](https://github.com/Hjdskes/libratbag/blob/3d44c6930592cab42c22862c713a10be22b894cb/data/README.md). To make it easy to test with these SVGs, modify lines [114 and 115](https://github.com/Hjdskes/piper/blob/c216782895d192266849e42cf4707a013a4745d2/piper/mousemap.py#L114..L115) to use absolute paths to your downloaded SVGs.

**Pull request merged**